services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightProducer.java

```diff
@@ -146,18 +217,39 @@ public void listActions(CallContext callContext, StreamListener<ActionType> stre
 
   @Override
   public void createPreparedStatement(
-    FlightSql.ActionCreatePreparedStatementRequest actionCreatePreparedStatementRequest,
+    ActionCreatePreparedStatementRequest actionCreatePreparedStatementRequest,
     CallContext callContext,
     StreamListener<Result> streamListener) {
-    throw CallStatus.UNIMPLEMENTED.withDescription("createPreparedStatement not supported.").toRuntimeException();
+    final FlightDescriptor flightDescriptor =
+      FlightDescriptor.command(actionCreatePreparedStatementRequest.getQuery().getBytes(StandardCharsets.UTF_8));
+
+    final UserSession session = getUserSessionFromCallContext(callContext);
+    final FlightPreparedStatement flightPreparedStatement = flightWorkManager
+      .createPreparedStatement(flightDescriptor, callContext::isCancelled, session);
+
+    flightPreparedStatementCache.put(flightPreparedStatement.getServerHandle(), flightPreparedStatement);
+
+    final ActionCreatePreparedStatementResult action = flightPreparedStatement.createAction();
+
+    streamListener.onNext(new Result(pack(action).toByteArray()));
+    streamListener.onCompleted();
+
   }
 
   @Override
   public void closePreparedStatement(
-    FlightSql.ActionClosePreparedStatementRequest actionClosePreparedStatementRequest,
+    ActionClosePreparedStatementRequest actionClosePreparedStatementRequest,
     CallContext callContext,
     StreamListener<Result> listener) {
-    throw CallStatus.UNIMPLEMENTED.withDescription("closePreparedStatement not supported.").toRuntimeException();
+    UserProtos.PreparedStatementHandle preparedStatementHandle;
+    try {
+      preparedStatementHandle =
+        UserProtos.PreparedStatementHandle.parseFrom(actionClosePreparedStatementRequest.getPreparedStatementHandle());
+    } catch (InvalidProtocolBufferException e) {
+      throw CallStatus.INVALID_ARGUMENT.withDescription("Invalid PreparedStatementHandle").toRuntimeException();
+    }
+
+    flightPreparedStatementCache.invalidate(preparedStatementHandle);
```

Comment:
Suggest you just put this in the try block.

Reply:
Fixed